plugins/ch341a/fu-ch341a-cfi-device.c

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-ch341a-cfi-device.h"
+#include "fu-ch341a-device.h"
+
+struct _FuCh341aCfiDevice {
+	FuCfiDevice parent_instance;
+};
+
+G_DEFINE_TYPE(FuCh341aCfiDevice, fu_ch341a_cfi_device, FU_TYPE_CFI_DEVICE)
+
+#define CH341A_PAYLOAD_SIZE 0x1A
+
+static gboolean
+fu_ch341a_cfi_device_chip_select(FuCfiDevice *self, gboolean value, GError **error)
+{
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(FU_DEVICE(self)));
+	return fu_ch341a_device_chip_select(proxy, value, error);
+}
+
+typedef struct {
+	guint8 mask;
+	guint8 value;
+} FuCh341aCfiDeviceHelper;
+
+static gboolean
+fu_ch341a_cfi_device_wait_for_status_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuCh341aCfiDeviceHelper *helper = (FuCh341aCfiDeviceHelper *)user_data;
+	FuCh341aCfiDevice *self = FU_CH341A_CFI_DEVICE(device);
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(device));
+	guint8 buf[2] = {0x0};
+	g_autoptr(FuDeviceLocker) cslocker = NULL;
+
+	/* enable chip */
+	cslocker = fu_cfi_device_chip_select_locker_new(FU_CFI_DEVICE(self), error);
+	if (cslocker == NULL)
+		return FALSE;
+	if (!fu_cfi_device_get_cmd(FU_CFI_DEVICE(self),
+				   FU_CFI_DEVICE_CMD_READ_STATUS,
+				   &buf[0],
+				   error))
+		return FALSE;
+	if (!fu_ch341a_device_spi_transfer(proxy, buf, sizeof(buf), error)) {
+		g_prefix_error(error, "failed to want to status: ");
+		return FALSE;
+	}
+	if ((buf[0x1] & helper->mask) != helper->value) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INTERNAL,
+			    "wanted 0x%x, got 0x%x",
+			    helper->value,
+			    buf[0x1] & helper->mask);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ch341a_cfi_device_wait_for_status(FuCh341aCfiDevice *self,
+				     guint8 mask,
+				     guint8 value,
+				     guint count,
+				     guint delay,
+				     GError **error)
+{
+	FuCh341aCfiDeviceHelper helper = {.mask = mask, .value = value};
+	return fu_device_retry_full(FU_DEVICE(self),
+				    fu_ch341a_cfi_device_wait_for_status_cb,
+				    count,
+				    delay,
+				    &helper,
+				    error);
+}
+
+static gboolean
+fu_ch341a_cfi_device_read_jedec(FuCh341aCfiDevice *self, GError **error)
+{
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(FU_DEVICE(self)));
+	guint8 buf[CH341A_PAYLOAD_SIZE] = {0x9F};
+	g_autoptr(FuDeviceLocker) cslocker = NULL;
+	g_autoptr(GString) flash_id = g_string_new(NULL);
+
+	/* enable chip */
+	cslocker = fu_cfi_device_chip_select_locker_new(FU_CFI_DEVICE(self), error);
+	if (cslocker == NULL)
+		return FALSE;
+
+	/* read JEDEC ID */
+	if (!fu_ch341a_device_spi_transfer(proxy, buf, sizeof(buf), error)) {
+		g_prefix_error(error, "failed to request JEDEC ID: ");
+		return FALSE;
+	}
+	if (buf[1] == 0xFF && buf[2] == 0xFF && buf[3] == 0xFF) {
```

Comment:
I get 00:00:00, does it make sense to consider that value as a error?

Reply:
Agreed, fixed.